Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -64,151 +60,156 @@ class SerializerTest : public ::testing::Test {
+    google::firestore::v1beta1::Value proto;
+    proto.set_integer_value(i);
+    return proto;
+  };
+
+  google::firestore::v1beta1::Value EncodeStringToProto(const std::string& s) {
+    google::firestore::v1beta1::Value proto;
+    proto.set_string_value(s);
+    return proto;
+  };
+
+ private:
+  void ExpectSerializationRoundTrip(
+      const FieldValue& model,
+      const google::firestore::v1beta1::Value& proto,
+      FieldValue::Type type) {
     EXPECT_EQ(type, model.type());
-    std::vector<uint8_t> actual_bytes;
-    Status status = serializer.EncodeFieldValue(model, &actual_bytes);
+    std::vector<uint8_t> bytes;
+    Status status = serializer.EncodeFieldValue(model, &bytes);
     EXPECT_TRUE(status.ok());
-    EXPECT_EQ(bytes, actual_bytes);
+    google::firestore::v1beta1::Value actual_proto;
+    bool ok = actual_proto.ParseFromArray(bytes.data(), bytes.size());
+    EXPECT_TRUE(ok);
+    EXPECT_TRUE(MessageDifferencer::Equals(proto, actual_proto));
+  }
+
+  void ExpectDeserializationRoundTrip(
+      const FieldValue& model,
+      const google::firestore::v1beta1::Value& proto,
+      FieldValue::Type type) {
+    size_t size = proto.ByteSizeLong();
+    std::vector<uint8_t> bytes(size);
+    bool status = proto.SerializeToArray(bytes.data(), size);
+    EXPECT_TRUE(status);
     FieldValue actual_model = serializer.DecodeFieldValue(bytes);
     EXPECT_EQ(type, actual_model.type());
     EXPECT_EQ(model, actual_model);
   }
 };
 
+// TODO(rsgowman): whoops! A previous commit performed approx s/Encodes/Writes/,
+// but should not have done so here. Change it back in this file.
+
 TEST_F(SerializerTest, WritesNullModelToBytes) {
   FieldValue model = FieldValue::NullValue();
 
-  // TEXT_FORMAT_PROTO: 'null_value: NULL_VALUE'
-  std::vector<uint8_t> bytes{0x58, 0x00};
-  ExpectRoundTrip(model, bytes, FieldValue::Type::Null);
+  google::firestore::v1beta1::Value proto;
+  proto.set_null_value(google::protobuf::NullValue::NULL_VALUE);
+
+  ExpectRoundTrip(model, proto, FieldValue::Type::Null);
 }
 
 TEST_F(SerializerTest, WritesBoolModelToBytes) {
-  struct TestCase {
-    bool value;
-    std::vector<uint8_t> bytes;
-  };
+  for (bool bool_value : {true, false}) {
+    FieldValue model = FieldValue::BooleanValue(bool_value);
 
-  std::vector<TestCase> cases{// TEXT_FORMAT_PROTO: 'boolean_value: true'
-                              {true, {0x08, 0x01}},
-                              // TEXT_FORMAT_PROTO: 'boolean_value: false'
-                              {false, {0x08, 0x00}}};
+    google::firestore::v1beta1::Value proto;
+    proto.set_boolean_value(bool_value);
 
-  for (const TestCase& test : cases) {
-    FieldValue model = FieldValue::BooleanValue(test.value);
-    ExpectRoundTrip(model, test.bytes, FieldValue::Type::Boolean);
+    ExpectRoundTrip(model, proto, FieldValue::Type::Boolean);
+  }
+}
+
+TEST_F(SerializerTest, WritesBoolModelToBytes_ALTERNATIVE) {
```

Comment:
Heads up: This test (obviously) is an alternative implementation of the above WritesBoolModelToBytes test, but uses the helper. Let me know which you prefer.

The obj-c assembled the proto manually (i.e. used the other variant), but then used the highler level "helper" functions in the map test. However, it was able to use portions of the "real" serializer to do so, whereas we can't here, so the helpers never really get proven.

Reply:
The helpers above are effectively identical to the in-line written ones so I'm fine reusing them here. I think for the simple cases like this it's easy enough to visually inspect that the low-level parts are constructing the right values.

IOW I'm fine with the alternative.